tianshou/data/buffer.py

```diff
@@ -520,13 +437,215 @@ def update_weight(
+        episode_rewards = []  # (len(buffer_ids), ...)
+        for batch_idx, env_id in enumerate(buffer_ids):
+            length, reward = self.buffers[env_id].add(**batch[batch_idx])
+            episode_lengths.append(length)
+            episode_rewards.append(reward)
+        return np.array(episode_lengths), np.array(episode_rewards)
+
+    def sample_index(self, batch_size: int) -> np.ndarray:
+        if batch_size < 0:
+            return np.array([], np.int)
+        if self._sample_avail and self.stack_num > 1:
+            all_indices = np.concatenate([
+                buf.sample_index(0) + buf.offset for buf in self.buffers])
+            if batch_size == 0:
+                return all_indices
+            else:
+                return np.random.choice(all_indices, batch_size)
+        if batch_size == 0:  # get all available indices
+            sample_num = np.zeros(self.buffer_num, np.int)
+        else:
+            buffer_lens = np.array([len(buf) for buf in self.buffers])
+            buffer_idx = np.random.choice(self.buffer_num, batch_size,
+                                          p=buffer_lens / buffer_lens.sum())
+            sample_num = np.bincount(buffer_idx, minlength=self.buffer_num)
+            # avoid batch_size > 0 and sample_num == 0 -> get child's all data
+            sample_num[sample_num == 0] = -1
+
+        return np.concatenate([
+            buf.sample_index(bsz) + buf.offset
+            for buf, bsz in zip(self.buffers, sample_num)
+        ])
+
+
+class CachedReplayBuffer(ReplayBuffers):
+    """CachedReplayBuffer contains a ReplayBuffers with given size and n \
+    cached buffers, cached_buffer_num * ReplayBuffer(size=max_episode_length).
+
+    The memory layout is: ``| main_buffer | cached_buffer[0] | cached_buffer[1]
+    | ... | cached_buffer[cached_buffer_num - 1]``.
+
+    The data is first stored in cached buffers. When the episode is
+    terminated, the data will move to the main buffer and the corresponding
+    cached buffer will be reset.
+
+    :param int size: the size of main buffer.
+    :param int cached_buffer_num: number of ReplayBuffer needs to be created
+        for cached buffer.
+    :param int max_episode_length: the maximum length of one episode, used in
+        each cached buffer's maxsize.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.data.ReplayBuffer` or
+        :class:`~tianshou.data.ReplayBuffers` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        size: int,
+        cached_buffer_num: int,
+        max_episode_length: int,
+        **kwargs: Any,
+    ) -> None:
+        assert cached_buffer_num > 0 and max_episode_length > 0
+        main_buffer = ReplayBuffer(size, **kwargs)
+        buffers = [main_buffer] + [ReplayBuffer(max_episode_length, **kwargs)
+                                   for _ in range(cached_buffer_num)]
+        super().__init__(buffer_list=buffers, **kwargs)
```

Comment:
suggestions:
self.main_buffer = main_buffer  (to replace buffers[0]) in line 648,etc.
self.cached_buffers = buffers[1:]

we use 'buffers ' to  initialise super(). but keep api clear to user